gateway/src/cron.rs

```diff
@@ -0,0 +1,427 @@
+
+                // return the current checkpoint's count as the current most submitted checkpoint
+                return Ok(new_count);
+            }
+
+            // the current submission is not part of the most submitted entries, need to check
+            // the most submitted entry to compare if the current submission is exceeding
+
+            // save to unwrap at the set cannot be empty
+            let most_submitted_hash = most_submitted_hashes.iter().next().unwrap();
+            let most_submitted_key = BytesKey::from(most_submitted_hash.as_slice());
+
+            // safe to unwrap as the hamt must contain the key
+            let most_submitted_count = hamt.get(&most_submitted_key)?.unwrap();
+
+            // current submission was not found in the most submitted checkpoints, the count gas is
+            // at least 1, new_count > *most_submitted_count will not happen
+            // if new_count < *most_submitted_count, we do nothing as the new count is not close to the most submitted
+            if new_count == *most_submitted_count {
+                most_submitted_hashes.insert(hash);
+            }
+
+            Ok(*most_submitted_count)
+        })
+    }
+
+    /// Checks if the submitter has already submitted the checkpoint. Currently used only in
+    /// tests, but can be used in prod as well.
+    #[cfg(test)]
+    fn has_submitted<BS: Blockstore>(
+        &self,
+        store: &BS,
+        submitter: &Address,
+    ) -> anyhow::Result<bool> {
+        let addr_byte_key = BytesKey::from(submitter.to_bytes());
+        let hamt = self.submitters.load(store)?;
+        Ok(hamt.contains_key(&addr_byte_key)?)
+    }
+
+    /// Checks if the checkpoint hash has already inserted in the store
+    #[cfg(test)]
+    fn has_checkpoint_inserted<BS: Blockstore>(
+        &self,
+        store: &BS,
+        hash: &HashOutput
+    ) -> anyhow::Result<bool> {
+        let hamt = self.submissions.load(store)?;
+        Ok(hamt.contains_key(&BytesKey::from(hash.as_slice()))?)
+    }
+
+    /// Checks if the checkpoint hash has already inserted in the store
+    #[cfg(test)]
+    fn get_submission_count<BS: Blockstore>(
+        &self,
+        store: &BS,
+        hash: &HashOutput
+    ) -> anyhow::Result<Option<u16>> {
+        let hamt = self.submission_counts.load(store)?;
+        let r = hamt.get(&BytesKey::from(hash.as_slice()))?;
+        Ok(r.cloned())
+    }
+}
+
+/// Compare the ordering of two storable messages.
+fn compare_top_down_msg(a: &StorableMsg, b: &StorableMsg) -> anyhow::Result<Ordering> {
+    let ordering = a.from.raw_addr()?.cmp(&b.from.raw_addr()?);
+    if ordering != Ordering::Equal {
+        return Ok(ordering);
+    }
+
+    let ordering = a.to.raw_addr()?.cmp(&b.to.raw_addr()?);
+    if ordering != Ordering::Equal {
+        return Ok(ordering);
+    }
+
+    Ok(a.nonce.cmp(&b.nonce))
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
These are really nice, I think is worth updating it after we simplify the logic.